### ISSUE TITLE
pihole -up needs to auto impliment pihole -g

### ISSUE THREAD

**Issue by Thomas499** [pi-hole/pi-hole#3960]
### Versions

- Pi-hole:  5.2.2
- AdminLTE:  5.2.2
- FTL:  5.3.4

### Platform

- OS and version:  

- Platform:  

### Expected behavior

When you run pihole -up I expect it to check for pihole updates and then automatically update gravity same as if you were to run pihole -g. 

### Actual behavior / bug

_A clear and concise description of what the bug is._

### Steps to reproduce

1. Run pihole -up
2. If there is an update it will install the update then automatically update gravity without user having to run pihole -g
3. If there is not an update then it does not automatically update gravity requiring the user to type pihole -g to update gravity after user already typed pihole -up
4. Regardless if there is an update to pihole available running pihole -up should go ahead and update the gravity list so that user only has to type in one command, and more importantly to implement consistency. 

## Debug Token

- URL:  

### Screenshots

_If applicable, add screenshots to help explain your problem._

### Additional context

_Add any other context about the problem here._

**Comment by technicalpyro** [comment]
this is not the default behaviour and i doubt it ever will be the two commands are separate for a reason 

`pihole -up` is for updating the application itself
`pihole -g` is to update gravity

**Comment by Thomas499** [comment]
But why does running pihole -up automatically update gravity if there is an update to pihole itself? 

The point is there needs to be consistency to what happens when user runs pihole -up regardless if there is an update available or not. 

I believe if user wants to update pihole then they would like to update everything about it including gravity in one single command. 

Of course keep pihole -g as an option in case user only wants to update gravity without the pihole framework. But when would user not want to update gravity when updating pihole? - And if there is an update to pihole it updates gravity automatically anyway (as is).

So my point is why not have running pihole -up automatically update gravity regardless if there is an update to pihole available?

**Comment by dschaper** [comment]
Because of the number of requests and the very limited resources we have as a free open-source project run by volunteers, we ask that you open all Feature Requests at our [Discourse Forum](https://discourse.pi-hole.net/c/feature-requests/8). 

Thank you for your understanding.

**Comment by DL6ER** [comment]
I will just comment on your question:

> So my point is why not have running pihole -up automatically update gravity regardless if there is an update to pihole available?

Because we know from a debug reports (generated via `pihole -d` and then uploaded) that an amazingly (or should I say "shockingly"?) high number of users have custom `cron` rules in place to run `pihole -up` automated and headless in 10 minute intervals. Running `pihole -g` at these intervals would lead to quite some issues.

**Comment by Thomas499** [comment]
Interesting. Thank you for taking the time to explain that.

Out of curiosity, if it is known that such a shockingly high number of people have custom cron rules in place to run pihole -up automated, then why not add a feature that lets the user schedule the cron rules within the pihole options? 

I think the options of once a day and once a month would suffice. I know I would like the auto update to happen without having to log in once a week to check. That would also do away with the problems you mentioned running pihole -g at those intervals would lead to because then the user wouldn't set custom cron rules to update every 10 minutes because it would be built into pihole already in a smart manner that only checks once a day instead of every 10 minutes. 

I know I would like to be able to set pihole to search for and download updates automatically at 3:00am every day as well as update gravity at that time every time. But I don't want to have to program custom cron rules i'd rather use the pi holes interface. 

I tried to figure out how to Discourse Forum to recommend this feature as dschaper requested, but for the life of me I can't figure out how to add a new feature request recommendation on that forum.

**Comment by jfb-pihole** [comment]
> I know I would like to be able to set pihole to search for and download updates automatically at 3:00am every da

This is highly discouraged.  We recommend a review of the release notes prior to any update of Pi-hole.  Some releases (V5.0 as an example) are not reversible, and you don't want to wake up in the morning to a new version that doesn't work for you and you can't go back since you weren't awake at 0230 to make the backup that was recommended.

**Comment by Thomas499** [comment]
`This is highly discouraged. `

As a user, my thoughts are I trust the developers to not push out an update that won't work for me and that the updates will improve my experience. I personally am not going to read the release notes before doing the update. I believe I am considered the typical user and that most other users do the same. 

 
As DL6ER stated:
`Because we know from a debug reports (generated via pihole -d and then uploaded) that an amazingly (or should I say "shockingly"?) high number of users have custom cron rules in place to run pihole -up automated and headless in 10 minute intervals`

It is a known fact that a shockingly high number of users want this feature so badly they are willing to actually spend the time to program custom cron rules to do it. That doesn't even account for the number of users that would fully appreciate the feature but don't want to spend the time learning how to program custom cron rules to make it happen. 

 If you want to provide the option that has been proven **many users want so badly they are going way out of their way to program the feature themself**, yet responsibly discourage it, then make the option not enabled by default, similar to how permit all origins is an option found in settings->DNS-> Interface listening behavior ->Listen on all interfaces, permit all origins but is not enabled by default. 

You can even go a step farther and list a foot note like how pihole did to explain why permit all origins is not recommended `Note that the last option should not be used on devices which are directly connected to the Internet.
`

**Comment by PromoFaux** [comment]
We will not be providing an auto update feature.

**Comment by DL6ER** [comment]
> As a user, my thoughts are I trust the developers to not push out an update that won't work for me and that the updates will improve my experience. I personally am not going to read the release notes before doing the update. I believe I am considered the typical user and that most other users do the same.

Just one more word to this: We can only test new features on very few systems we have available to ourselves. We are not earning any money by selling any kind of licenses or whatsoever so our testing gear is certainly limited. You can typically assume that Pi-hole works on Raspberry Pis, Debian, Ubuntu, Fedora and CentOS. Given all the possible configurations that are possible with Pi-hole these days (query type dependent regex blocking, DHCP server, custom user configuration to achieve some special VLAN handling, etc.), it is just impossible to cover everything. Let alone old devices that were not updated for 5+ years.

For the typical user - running a fairly up-to-date Raspberry Pi - an automated update should always succeed. However, more an more users are installing Pi-hole on devices for which it was never meant. You can find everything on discourse.pi-hole.net ranging from smart-home equipment like smart-plugs to installing it natively on NAS devices or even on Smart-TVs. Furthermore, there is a vast amount of heavily tweaked operating systems out there to target other SoC boards like NanoPis, OrangePis, etc. Then there is also a modified installer inside the popular operating system DietPi and more.

Even when such a feature would be disabled by default, I'm afraid too many would switch it on because they do the same as you:

> I trust the developers to not push out an update that won't work for me

This may (read as: *is often enough*) just not true for heavily tweaked and not directly supported devices. In the end, it would break a lot more than it would be good for. If your Pi-hole is running as intended, just leave it as it is. Updating without using the new features (because you are not even aware of their existence) doesn't bring you anything, typically. Yes, there may be security improvements, however, Pi-hole is performing very good in this regard in the wild and you should *always* have a solid firewall in front of your Pi-hole such that this isn't much of a headache.

TL;DR: What @PromoFaux said above.

**Comment by dschaper** [comment]
We use Semantic Versioning. This means that there will be incompatible updates.

>Given a version number MAJOR.MINOR.PATCH, increment the:
>
>    MAJOR version when you make incompatible API changes,
>    MINOR version when you add functionality in a backwards compatible manner, and
>    PATCH version when you make backwards compatible bug fixes.
>
> Additional labels for pre-release and build metadata are available as extensions to the MAJOR.MINOR.PATCH format.

Pi-hole is the DNS server for your network. It's a vital service, if it doesn't work, your network effectively doesn't work. And yes, there are a lot of people that have automatic update scripts. They soon disable them the first time an update takes down their networks. So hear me now and believe me later, automatic updates will break your network.
